src/Agent.PluginCore/TaskPlugin.cs

```diff
@@ -0,0 +1,224 @@
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.OAuth;
+using Microsoft.VisualStudio.Services.WebApi;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Pipelines = Microsoft.TeamFoundation.DistributedTask.Pipelines;
+
+namespace Agent.PluginCore
+{
+    public interface IAgentTaskPlugin
+    {
+        Guid Id { get; }
+        string Version { get; }
+        string FriendlyName { get; }
+        string Description { get; }
+        string HelpMarkDown { get; }
+        string Author { get; }
+        TaskInputDefinition[] Inputs { get; }
+        HashSet<string> Stages { get; }
+        Task RunAsync(AgentTaskPluginExecutionContext executionContext, CancellationToken token);
+    }
+
+    public class AgentTaskPluginExecutionContext
+    {
+        public AgentTaskPluginExecutionContext()
+        {
+            this.Endpoints = new List<ServiceEndpoint>();
+            this.Inputs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            this.Repositories = new List<Pipelines.RepositoryResource>();
+            this.TaskVariables = new Dictionary<string, VariableValue>(StringComparer.OrdinalIgnoreCase);
+            this.Variables = new Dictionary<string, VariableValue>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public string Stage { get; set; }
+        public List<ServiceEndpoint> Endpoints { get; set; }
+        public List<Pipelines.RepositoryResource> Repositories { get; set; }
+        public Dictionary<string, VariableValue> Variables { get; set; }
+        public Dictionary<string, VariableValue> TaskVariables { get; set; }
+        public Dictionary<string, string> Inputs { get; set; }
+
+        public string GetInput(string name, bool required = false)
+        {
+            string value = null;
+            if (this.Inputs.ContainsKey(name))
+            {
+                value = this.Inputs[name];
+            }
+
+            if (string.IsNullOrEmpty(value) && required)
+            {
+                throw new ArgumentNullException(name);
+            }
+
+            return value;
+        }
+
+        public void Fail(string message)
+        {
+            if (!string.IsNullOrEmpty(message))
+            {
+                Error(message);
+            }
+
+            Console.WriteLine($"##vso[task.complete result=Failed;]");
+        }
+
+        public void Debug(string message)
+        {
+            Console.WriteLine($"##vso[task.debug]{message}");
+        }
+
+        public void Error(string message)
```

Comment:
do we ever use this? i would just remove this overload and rename fail to error. we can always add an overload later.

Reply:
error will output ##vso[task.setresult result=failed;] by default

Reply:
fixed.